### PR TITLE
Modify Travis Jobs

### PR DESCRIPTION
**Pull request by dnoliver**
Use Docker and docker-compose to run Travis Jobs.

Similar to PR https://github.com/intel-iot-devkit/mraa/pull/744

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  libmraa1 libmraa-dev mraa-tools python-mraa python3-mraa libmraa-java \
+  # Docs Build Dependencies
+  python-sphinx doxygen graphviz \
+  # Python Build Dependencies
+  python python-dev python3 python3-dev \
+  # Java Build Dependencies
+  default-jre default-jdk \
+  # Sensor Specific Build Dependencies
+  libmodbus-dev libjpeg-dev
+  # TODO: install libbacnet
```
Comment (dnoliver) [inline Dockerfile]: Posted an issue to get proper install steps for libbacnet #563

**Comment by bjbeare** [comment]
I created the docker and ran into an issue.

Installing /usr/share/swig/3.0.10/std/std_wsstream.i
Installing /usr/share/swig/3.0.10/std/std_wstreambuf.i
Installing /usr/share/swig/3.0.10/std/std_wstring.i
Installing /usr/share/swig/3.0.10/std/std_carray.swg
make[1]: Entering directory '/swig-3.0.10/CCache'
Installing ccache-swig
Installing /usr/bin/ccache-swig
/usr/bin/install -c -d /usr/bin
/usr/bin/install -c -m 755 ccache-swig /usr/bin/`echo ccache-swig | sed 's,x,x,'`
make[1]: Leaving directory '/swig-3.0.10/CCache'
Installation complete
 ---> b067e547a7cd
Removing intermediate container 0ef30ed21c2e
Step 7 : RUN wget -q -O - https://raw.githubusercontent.com/creationix/nvm/v0.33.2/install.sh | bash
 ---> Running in 7251ceb2b3c6
=> Downloading nvm from git to '/root/.nvm'
=> Cloning into '/root/.nvm'...
* (HEAD detached at v0.33.2)
  master
=> Compressing and cleaning up git repository

=> Appending nvm source string to /root/.bashrc
=> Appending bash_completion source string to /root/.bashrc
=> Close and reopen your terminal to start using nvm or run the following to use it now:

export NVM_DIR="$HOME/.nvm"
[ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"  # This loads nvm
[ -s "$NVM_DIR/bash_completion" ] && \. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion
 ---> 527b4ea9a0fb
Removing intermediate container 7251ceb2b3c6
Step 8 : ARG NODE_VERSION
 ---> Running in 0da2854c411d
 ---> 2f58f7b6a8e5
Removing intermediate container 0da2854c411d
Step 9 : ENV NVM_DIR /root/.nvm
 ---> Running in a6680c18da39
 ---> aadc23e546f7
Removing intermediate container a6680c18da39
Step 10 : RUN . $NVM_DIR/nvm.sh && nvm install $NODE_VERSION && nvm use $NODE_VERSION &&     npm install -g node-gyp && node-gyp install
 ---> Running in 721ad2f85e9f
No .nvmrc file found

Node Version Manager

Note: <version> refers to any version-like string nvm understands. This includes:
  - full or partial version numbers, starting with an optional "v" (0.10, v0.1.2, v1)
  - default (built-in) aliases: node, stable, unstable, iojs, system
  - custom aliases you define with `nvm alias foo`

 Any options that produce colorized output should respect the `--no-colors` option.

Usage:
  nvm --help                                Show this message
  nvm --version                             Print out the latest released version of nvm
  nvm install [-s] <version>                Download and install a <version>, [-s] from source. Uses .nvmrc if available
    --reinstall-packages-from=<version>     When installing, reinstall packages installed in <node|iojs|node version number>
    --lts                                   When installing, only select from LTS (long-term support) versions
    --lts=<LTS name>                        When installing, only select from versions for a specific LTS line
  nvm uninstall <version>                   Uninstall a version
  nvm uninstall --lts                       Uninstall using automatic LTS (long-term support) alias `lts/*`, if available.
  nvm uninstall --lts=<LTS name>            Uninstall using automatic alias for provided LTS line, if available.
  nvm use [--silent] <version>              Modify PATH to use <version>. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm exec [--silent] <version> [<command>] Run <command> on <version>. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm run [--silent] <version> [<args>]     Run `node` on <version> with <args> as arguments. Uses .nvmrc if available
    --lts                                   Uses automatic LTS (long-term support) alias `lts/*`, if available.
    --lts=<LTS name>                        Uses automatic alias for provided LTS line, if available.
  nvm current                               Display currently activated version
  nvm ls                                    List installed versions
  nvm ls <version>                          List versions matching a given <version>
  nvm ls-remote                             List remote versions available for install
    --lts                                   When listing, only show LTS (long-term support) versions
  nvm ls-remote <version>                   List remote versions available for install, matching a given <version>
    --lts                                   When listing, only show LTS (long-term support) versions
    --lts=<LTS name>                        When listing, only show versions for a specific LTS line
  nvm version <version>                     Resolve the given description to a single local version
  nvm version-remote <version>              Resolve the given description to a single remote version
    --lts                                   When listing, only select from LTS (long-term support) versions
    --lts=<LTS name>                        When listing, only select from versions for a specific LTS line
  nvm deactivate                            Undo effects of `nvm` on current shell
  nvm alias [<pattern>]                     Show all aliases beginning with <pattern>
  nvm alias <name> <version>                Set an alias named <name> pointing to <version>
  nvm unalias <name>                        Deletes the alias named <name>
  nvm reinstall-packages <version>          Reinstall global `npm` packages contained in <version> to current version
  nvm unload                                Unload `nvm` from shell
  nvm which [<version>]                     Display path to installed node version. Uses .nvmrc if available
  nvm cache dir                             Display path to the cache directory for nvm
  nvm cache clear                           Empty cache directory for nvm

Example:
  nvm install v0.10.32                  Install a specific version number
  nvm use 0.10                          Use the latest available 0.10.x release
  nvm run 0.10.32 app.js                Run app.js using node v0.10.32
  nvm exec 0.10.32 node app.js          Run `node app.js` with the PATH pointing to node v0.10.32
  nvm alias default 0.10.32             Set default node version on a shell

Note:
  to remove, delete, or uninstall nvm - just remove the `$NVM_DIR` folder (usually `~/.nvm`)

The command '/bin/sh -c . $NVM_DIR/nvm.sh && nvm install $NODE_VERSION && nvm use $NODE_VERSION &&     npm install -g node-gyp && node-gyp install' returned a non-zero code: 127

**Comment by dnoliver** [comment]
@bjbeare how do you executed the command? and which version of docker and docker-compose you are using?

**Comment by bjbeare** [comment]
Ubuntu 16.04 - installed from apt-get.
(it was at home... so I don't have access to it just now).

**Comment by dnoliver** [comment]
Commit https://github.com/intel-iot-devkit/upm/pull/564/commits/c70d06bfbf93ea9f882fb19ff4f6953d68475907 install optional dependencies. libbacnet install is a workaround until we have a better way to do it. 

#563

**Comment by dnoliver** [comment]
Travis Jobs still failing, will need help!

Node4 and Node5 Examples test fail
```
Missing examples:
a110x-intr: a110x-intr.js
hcsr04: hcsr04.js
i2clcd-ssd1308-oled: ssd1308-oled.js
i2clcd-ssd1327-oled: ssd1327-oled.js
lcm1602-i2c: lcm1602-i2c.js
lol: lol.js
lsm303dlh: lsm303dlh.js
max44000: max44000.js
mma7455: mma7455.js
mpl3115a2: mpl3115a2.js
nrf24l01-receiver: nrf24l01-receiver.js
nrf24l01-transmitter: nrf24l01-transmitter.js
pulsensor: pulsensor.js
rpr220-intr: rpr220-intr.js
```

Java build Fail because of OpenZWave lib
```
[ 35%] Linking CXX executable xbee-example-cxx
Scanning dependencies of target i2clcd-ssd1306-oled-example-cxx
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeProductId(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Options::AddOptionBool(std::string const&, bool)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Options::AddOptionString(std::string const&, std::string const&, bool)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetValueAsString(OpenZWave::ValueID const&, std::string*)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeManufacturerId(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::SetValueLabel(OpenZWave::ValueID const&, std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::AddDriver(std::string const&, OpenZWave::Driver::ControllerInterface const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::RemoveDriver(std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::SetValueHelp(OpenZWave::ValueID const&, std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeProductType(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeProductName(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Options::Create(std::string const&, std::string const&, std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeQueryStage(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetNodeType(unsigned int, unsigned char)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::SetValueUnits(OpenZWave::ValueID const&, std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetValueHelp(OpenZWave::ValueID const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::SetValue(OpenZWave::ValueID const&, std::string const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetValueUnits(OpenZWave::ValueID const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Manager::GetValueLabel(OpenZWave::ValueID const&)'
../../src/ozw/libupm-ozw.so.1.3.0: undefined reference to `OpenZWave::Options::AddOptionInt(std::string const&, int)'
collect2: error: ld returned 1 exit status
examples/c++/CMakeFiles/ozw-aeotecss6-example-cxx.dir/build.make:95: recipe for target 'examples/c++/ozw-aeotecss6-example-cxx' failed
make[2]: *** [examples/c++/ozw-aeotecss6-example-cxx] Error 1
CMakeFiles/Makefile2:35699: recipe for target 'examples/c++/CMakeFiles/ozw-aeotecss6-example-cxx.dir/all' failed
make[1]: *** [examples/c++/CMakeFiles/ozw-aeotecss6-example-cxx.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Python2 and Python3 module load test fail (incomplete stack trace)
```
Failed to load 221 modules:
/usr/src/app/build/src/a110x/python3.5/pyupm_a110x.py: No module named '_pyupm_a110x'
/usr/src/app/build/src/abp/python3.5/pyupm_abp.py: No module named '_pyupm_abp'
/usr/src/app/build/src/ad8232/python3.5/pyupm_ad8232.py: No module named '_pyupm_ad8232'
/usr/src/app/build/src/adafruitms1438/python3.5/pyupm_adafruitms1438.py: No module named '_pyupm_adafruitms1438'
/usr/src/app/build/src/adafruitss/python3.5/pyupm_adafruitss.py: No module named '_pyupm_adafruitss'
/usr/src/app/build/src/adc121c021/python3.5/pyupm_adc121c021.py: No module named '_pyupm_adc121c021'
/usr/src/app/build/src/adis16448/python3.5/pyupm_adis16448.py: No module named '_pyupm_adis16448'
/usr/src/app/build/src/ads1x15/python3.5/pyupm_ads1x15.py: No module named '_pyupm_ads1x15'
/usr/src/app/build/src/adxl335/python3.5/pyupm_adxl335.py: No module named '_pyupm_adxl335'
/usr/src/app/build/src/adxl345/python3.5/pyupm_adxl345.py: No module named '_pyupm_adxl345'
/usr/src/app/build/src/adxrs610/python3.5/pyupm_adxrs610.py: No module named '_pyupm_adxrs610'
/usr/src/app/build/src/am2315/python3.5/pyupm_am2315.py: No module named '_pyupm_am2315'
/usr/src/app/build/src/apa102/python3.5/pyupm_apa102.py: No module named '_pyupm_apa102'
```

**Comment by Propanu** [comment]
Seems like OZW hardcodes the `lib` or `lib64` suffix, and `/usr/local/lib64` is non-standard on Ubuntu. So you'll probably need to set an `ENV LD_LIBRARY_PATH=/usr/local/lib64` when you install OZW, that way, repeated calls to `ldconfig` will still pick it up.

File: docker-compose.yaml [hunk 2]
```diff
@@ -10,21 +10,21 @@ services:
...
-        - CC=clang-3.8
-        - CXX=clang++-3.8
-        - NODE_VERSION=v4.1.2
-        - WERROR=ON
+        - BUILDDOC=${BUILDDOC:-OFF}
+        - BUILDCPP=${BUILDCPP:-ON}
+        - BUILDFTI=${BUILDFTI:-ON}
+        - BUILDSWIGPYTHON=${BUILDSWIGPYTHON:-OFF}
+        - BUILDSWIGJAVA=${BUILDSWIGJAVA:-OFF}
+        - BUILDSWIGNODE=${BUILDSWIGJAVA:-OFF}
```
Comment (Propanu) [inline docker-compose.yaml]: Typo.

**Comment by dnoliver** [comment]
Apparently python build is broken because swig 3.0.10 has changed the way how packages are imported (https://github.com/swig/swig/issues/769). In the previous travis job, it was using swig3.0.7. I will change the swig version to test this,, but I think we should take it into account for the future. Maybe creating an issue that this will not work with swig > 3.8?

**Comment by Propanu** [comment]
Merged, thanks @dnoliver!!

**Pull request closed without merging**